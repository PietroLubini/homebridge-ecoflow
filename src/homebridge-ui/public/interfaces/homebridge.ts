export interface IHomebridge {
  getPluginConfig: () => Promise<PluginConfig[]>;
  getPluginConfigSchema: () => Promise<PluginConfigSchema>;
  updatePluginConfig: (changes: PluginConfig[]) => Promise<void>;
  createForm: (schema: { schema: PluginConfigSchemaDevicesItems }, configuration: PluginDeviceConfig) => IForm;
}

export interface PluginConfig {
  name: string;
  platform?: string;
  devices: PluginDeviceConfig[];
}

export interface PluginDeviceConfig {
  name: string;
  model: string;
  [name: string]: string | undefined;
}

export interface PluginConfigSchema {
  schema: {
    properties: {
      name: PluginConfigSchemaObject;
      devices: PluginConfigSchemaDevices;
    };
  };
}

export interface PluginConfigSchemaObject {
  title: string;
  description?: string;
  default?: string;
  required?: boolean;
}

export interface PluginConfigSchemaEnum extends PluginConfigSchemaObject {
  enum: string[];
}

export interface PluginConfigSchemaDevices {
  title: string;
  items: PluginConfigSchemaDevicesItems;
}

export interface PluginConfigSchemaDevicesItems {
  properties: PluginConfigSchemaDevice;
}

export interface PluginConfigSchemaDevice {
  name: PluginConfigSchemaObject;
  model?: PluginConfigSchemaEnum;
  [name: string]: PluginConfigSchemaObject | undefined;
}

export type FormOnChangeCallback = (newConfiguration: PluginDeviceConfig) => Promise<void>;

export interface IForm {
  onChange: (callback: FormOnChangeCallback) => void;
  end: () => void;
}
