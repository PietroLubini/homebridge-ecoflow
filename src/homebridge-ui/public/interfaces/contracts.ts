import { IHomebridge, PluginConfig, PluginConfigSchema } from '../interfaces/homebridge';

export interface IContext {
  homebridgeProvider: IHomebridge;
  configSchema: PluginConfigSchema;
  configuration: PluginConfig;
}

export interface IDeviceContext extends IContext {
  hideDeviceSettingsPerModel: Record<string, string[]>;
}

export type ChangedCallbackHandler = (value: string) => void;

export interface IRenderer {
  render(context: IContext): void;
}
