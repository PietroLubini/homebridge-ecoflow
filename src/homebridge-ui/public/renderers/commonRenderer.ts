import { ChangedCallbackHandler } from '../interfaces/contracts';
import { PluginConfigSchemaEnum, PluginConfigSchemaObject } from '../interfaces/homebridge';

export class CommonRenderer {
  public renderTextBox(
    $parent: JQuery,
    id: string,
    schemaProperty: PluginConfigSchemaObject,
    value: string,
    onChangeCallback: ChangedCallbackHandler
  ): void {
    const template = `
    <div class="form-group">
      <label for="device{id}" class="{requiredClass}">{title}</label>
      <input type="text" class="form-control" id="device{id}" placeholder="{description}" required="{required}" />
    </div>`;

    const html = template
      .replace(/{id}/g, id)
      .replace(/{title}/g, schemaProperty.title)
      .replace(/{description}/g, schemaProperty.description ?? '')
      .replace(/{requiredClass}/g, schemaProperty.required ? 'required-label' : '')
      .replace(/{required}/g, (schemaProperty.required ?? false).toString());
    $parent.append(html);

    const $control = $(`#device${id}`, $parent);
    $control.val(value);
    $control.on('change', () => onChangeCallback($control.val()?.toString() || ''));
  }

  public renderDropDown(
    $parent: JQuery,
    id: string,
    schemaProperty: PluginConfigSchemaEnum,
    value: string,
    onChangeCallback: ChangedCallbackHandler
  ): void {
    const template = `
    <div class="form-group">
      <label for="device{id}" class="{requiredClass}">{title}</label>
      <select class="form-control" id="device{id}" required="{required}"></select>
    </div>`;
    const html = template
      .replace(/{id}/g, id)
      .replace(/{title}/g, schemaProperty.title)
      .replace(/{requiredClass}/g, schemaProperty.required ? 'required-label' : '')
      .replace(/{required}/g, (schemaProperty.required ?? false).toString());
    $parent.append(html);

    const $control = $(`#device${id}`, $parent);
    schemaProperty.enum.forEach((enumValue: string) => {
      const $option = $('<option>').val(enumValue).text(enumValue);
      $control.append($option);
    });
    if (value !== undefined) {
      $control.val(value);
    }
    $control.on('change', () => onChangeCallback($control.val()?.toString() || ''));
  }
}
