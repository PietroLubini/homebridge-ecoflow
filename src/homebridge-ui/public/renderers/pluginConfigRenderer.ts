import { IContext, IRenderer } from '../interfaces/contracts';

export class PluginConfigRenderer {
  constructor(private readonly renderers: IRenderer[]) {}

  public render(context: IContext): void {
    if (context.configuration === undefined) {
      context.configuration = {
        name: undefined as unknown as string,
        devices: [],
      };
    }
    this.renderers.forEach(renderer => renderer.render(context));
  }
}
