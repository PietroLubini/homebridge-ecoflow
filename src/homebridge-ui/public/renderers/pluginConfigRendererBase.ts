import { IContext, IRenderer } from '../interfaces/contracts';
import { ComponentRenderer } from './componentRenderer';

export abstract class PluginConfigRendererBase implements IRenderer {
  constructor(protected readonly componentRenderer: ComponentRenderer) {}

  public abstract render(context: IContext): void;

  protected async updatePluginConfig(context: IContext): Promise<void> {
    const changes = this.clone(context.configuration);
    delete changes.platform;
    await context.homebridgeProvider.updatePluginConfig([changes]);
  }

  protected clone<TObject>(obj: TObject): TObject {
    return JSON.parse(JSON.stringify(obj));
  }
}
