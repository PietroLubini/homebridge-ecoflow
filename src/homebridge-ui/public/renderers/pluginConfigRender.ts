import { IContext, IDeviceContext } from '../interfaces/contracts';
import {
  IForm,
  IHomebridge,
  PluginConfig,
  PluginConfigSchema,
  PluginConfigSchemaDevicesItems,
  PluginConfigSchemaEnum,
  PluginDeviceConfig,
} from '../interfaces/homebridge';
import { CommonRenderer } from './commonRenderer';

export class PluginConfigRenderer {
  private form: IForm | undefined;

  constructor(private readonly commonRenderer: CommonRenderer) {}

  public render(
    homebridgeProvider: IHomebridge,
    configuration: PluginConfig,
    configSchema: PluginConfigSchema,
    hideDeviceSettingsPerModel: Record<string, string[]>
  ): void {
    this.renderGeneralSettings({ homebridgeProvider, configuration, configSchema });
    this.renderDevicesSettings({ homebridgeProvider, configSchema, configuration, hideDeviceSettingsPerModel });
  }

  private renderGeneralSettings(context: IContext): void {
    this.commonRenderer.renderTextBox(
      $('#generalSettings'),
      'name',
      context.configSchema.schema.properties.name,
      context.configuration.name,
      async newValue => {
        context.configuration.name = newValue;
        await this.updatePluginConfig(context);
      }
    );
  }

  private renderDevicesSettings(context: IDeviceContext, activeDeviceIndex?: number): void {
    $('#devicesLabel').text(context.configSchema.schema.properties['devices'].title);
    const activeIndex = !activeDeviceIndex || activeDeviceIndex < 0 ? 0 : activeDeviceIndex;
    this.renderDeviceTabs(context, activeIndex);
  }

  private renderDeviceTabs(context: IDeviceContext, activeIndex: number): void {
    const { $tabs, $tabPanels } = this.clearDeviceTabs();

    const tabTemplate = `
      <li class="nav-item">
        <a class="nav-link {activeClass}" data-toggle="tab" href="#deviceTabPanel{index}" id="deviceTab{index}">{name}</a>
      </li>
    `;

    context.configuration.devices.forEach((deviceConfiguration: PluginDeviceConfig, index: number) => {
      const activeTabClass = index === activeIndex ? 'active' : '';

      const tabHtml = tabTemplate
        .replace(/{activeClass}/g, activeTabClass)
        .replace(/{index}/g, index.toString())
        .replace(/{name}/g, deviceConfiguration.name ?? `Device${index + 1}`);

      $tabs.append(tabHtml);

      this.renderDeviceTabPanel(
        context,
        deviceConfiguration,
        index,
        activeIndex,
        $(`#deviceTab${index}`, $tabs),
        $tabPanels
      );
    });

    this.renderAddNewDeviceTab($tabs, tabTemplate, context);
  }

  private clearDeviceTabs(): { $tabs: JQuery; $tabPanels: JQuery } {
    const $tabs = $('#devicesTabs');
    const $tabPanels = $('#devicesPanels');
    $tabs.empty();
    $tabPanels.empty();
    this.form?.end();
    return { $tabs, $tabPanels };
  }

  private renderAddNewDeviceTab($parent: JQuery, template: string, context: IDeviceContext): void {
    const html = template
      .replace(/{activeClass}/g, '')
      .replace(/{index}/g, 'Add')
      .replace(/{name}/g, 'Add Device');
    $parent.append(html);
    const $addButton = $('#deviceTabAdd', $parent);
    $addButton.on('click', () => {
      context.configuration.devices.push(this.getDefaultDeviceConfiguration(context.configSchema));
      this.renderDevicesSettings(context, context.configuration.devices.length - 1);
    });
  }

  private renderDeviceTabPanel(
    context: IDeviceContext,
    deviceConfiguration: PluginDeviceConfig,
    index: number,
    activeIndex: number,
    $tab: JQuery,
    $tabPanels: JQuery
  ): void {
    const tabPanelTemplate = `
    <div class="tab-pane container fade {activeClass} card card-body list-group-item" id="deviceTabPanel{index}">
      <button type="button" class="close pull-right" id="deviceTabPanelClose{index}">
        <span>Ã—</span>
        <span class="sr-only">Close</span>
      </button>
    </div>
  `;

    const activeTabPanelClass = index === activeIndex ? 'in show active' : '';
    const tabPanelHtml = tabPanelTemplate
      .replace(/{activeClass}/g, activeTabPanelClass)
      .replace(/{index}/g, index.toString());

    $tabPanels.append(tabPanelHtml);
    const $tabPanel = $(`#deviceTabPanel${index}`, $tabPanels);

    const deviceSchema = this.clone(context.configSchema.schema.properties.devices.items);
    const modelSchema = deviceSchema.properties.model;
    this.renderModelDropDown($tabPanel, index, context, modelSchema!, deviceConfiguration);
    this.renderRemoveButton($tabPanel, index, context);
    this.renderForm($tab, index, activeIndex, context, deviceSchema, deviceConfiguration);
  }

  private renderModelDropDown(
    $parent: JQuery,
    index: number,
    context: IDeviceContext,
    modelSchema: PluginConfigSchemaEnum,
    deviceConfiguration: PluginDeviceConfig
  ) {
    this.commonRenderer.renderDropDown($parent, 'model' + index, modelSchema, deviceConfiguration.model, newValue => {
      deviceConfiguration.model = newValue;
      this.renderDevicesSettings(context, index);
    });
  }

  private renderRemoveButton($parent: JQuery, index: number, context: IDeviceContext) {
    const $removeButton = $(`#deviceTabPanelClose${index}`, $parent);
    $removeButton.on('click', () => {
      context.configuration.devices.splice(index, 1);
      this.renderDevicesSettings(context, index - 1);
    });
  }

  private renderForm(
    $tab: JQuery,
    index: number,
    activeIndex: number,
    context: IDeviceContext,
    deviceSchema: PluginConfigSchemaDevicesItems,
    deviceConfiguration: PluginDeviceConfig
  ): void {
    delete deviceSchema.properties.model;
    context.hideDeviceSettingsPerModel[deviceConfiguration.model].forEach(propertyName => {
      delete deviceSchema.properties[propertyName];
    });

    $tab.on('click', () => {
      this.form = context.homebridgeProvider.createForm({ schema: deviceSchema }, deviceConfiguration);
      this.form.onChange(async newDeviceConfiguration => {
        const oldName = context.configuration.devices[index].name;
        this.applyChanges(context.configuration.devices[index], newDeviceConfiguration, ['model']);
        await this.updatePluginConfig(context);

        if (newDeviceConfiguration.name !== oldName) {
          $tab.text(newDeviceConfiguration.name);
        }
      });
    });
    if (index === activeIndex) {
      $tab.trigger('click');
    }
  }

  private getDefaultDeviceConfiguration(configSchema: PluginConfigSchema): PluginDeviceConfig {
    const result = {} as PluginDeviceConfig;
    const properties = configSchema.schema.properties.devices.items.properties;
    Object.keys(properties)
      .filter(propertyName => !!properties[propertyName]?.default)
      .forEach(propertyName => {
        result[propertyName] = properties[propertyName]?.default;
      });
    return result;
  }

  private async updatePluginConfig(context: IContext): Promise<void> {
    const changes = this.clone(context.configuration);
    delete changes.platform;
    await context.homebridgeProvider.updatePluginConfig([changes]);
  }

  private clone<TObject>(obj: TObject): TObject {
    return JSON.parse(JSON.stringify(obj));
  }

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  private applyChanges(target: any, source: any, readOnlyProperties: string[]): void {
    Object.assign(target, source);
    for (const key in target) {
      if (source[key] === undefined && !readOnlyProperties.includes(key)) {
        delete target[key];
      }
    }
  }
}
