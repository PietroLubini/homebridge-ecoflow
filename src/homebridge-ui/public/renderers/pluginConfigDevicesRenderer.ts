import { IContext, IDeviceContext } from '../interfaces/contracts';
import {
  IForm,
  PluginConfigSchema,
  PluginConfigSchemaDevicesItems,
  PluginConfigSchemaEnum,
  PluginDeviceConfig,
} from '../interfaces/homebridge';
import { ComponentRenderer } from './componentRenderer';
import { PluginConfigRendererBase } from './pluginConfigRendererBase';

export class PluginConfigDevicesRenderer extends PluginConfigRendererBase {
  private form: IForm | undefined;
  private readonly $tabs: JQuery;
  private readonly $tabPanels: JQuery;
  private readonly _hideDeviceSettingsPerModel: Record<string, string[]>;

  constructor(componentRenderer: ComponentRenderer) {
    super(componentRenderer);
    this.$tabs = $('#devicesTabs');
    this.$tabPanels = $('#devicesPanels');
    this._hideDeviceSettingsPerModel = {
      'Delta 2': ['powerStream', 'powerOcean', 'outlet'],
      'Delta 2 Max': ['powerStream', 'powerOcean', 'outlet'],
      'Delta Pro': ['powerStream', 'powerOcean', 'outlet'],
      'Delta Pro 3': ['powerStream', 'powerOcean', 'outlet'],
      'Delta Pro Ultra': ['powerStream', 'powerOcean', 'outlet'],
      PowerStream: ['battery', 'powerOcean', 'outlet'],
      PowerOcean: ['battery', 'powerStream', 'outlet'],
      'Smart Plug': ['battery', 'powerStream', 'powerOcean'],
      Glacier: ['powerStream', 'powerOcean', 'outlet'],
    };
  }

  public get hideDeviceSettingsPerModel(): Record<string, string[]> {
    return this._hideDeviceSettingsPerModel;
  }

  public render(context: IContext): void {
    this.renderDevicesSettings({
      homebridgeProvider: context.homebridgeProvider,
      configSchema: context.configSchema,
      configuration: context.configuration,
      hideDeviceSettingsPerModel: this.hideDeviceSettingsPerModel,
    });
  }

  private renderDevicesSettings(context: IDeviceContext, activeDeviceIndex?: number): void {
    $('#devicesLabel').text(context.configSchema.schema.properties['devices'].title);
    const activeIndex = !activeDeviceIndex || activeDeviceIndex < 0 ? 0 : activeDeviceIndex;
    this.renderDeviceTabs(context, activeIndex);
  }

  private renderDeviceTabs(context: IDeviceContext, activeIndex: number): void {
    this.clearDeviceTabs();

    const tabTemplate = `
      <li class="nav-item">
        <a class="nav-link {activeClass}" data-toggle="tab" href="#deviceTabPanel{index}" id="deviceTab{index}">{name}</a>
      </li>
    `;

    context.configuration.devices.forEach((deviceConfiguration: PluginDeviceConfig, index: number) => {
      const activeTabClass = index === activeIndex ? 'active' : '';

      const tabHtml = tabTemplate
        .replace(/{activeClass}/g, activeTabClass)
        .replace(/{index}/g, index.toString())
        .replace(/{name}/g, deviceConfiguration.name ?? `Device${index + 1}`);

      this.$tabs.append(tabHtml);

      this.renderDeviceTabPanel(
        context,
        deviceConfiguration,
        index,
        activeIndex,
        this.$tabs.find(`#deviceTab${index}`)
      );
    });

    this.renderAddNewDeviceTab(tabTemplate, context);
  }

  private clearDeviceTabs(): void {
    this.$tabs.empty();
    this.$tabPanels.empty();
    this.form?.end();
  }

  private renderAddNewDeviceTab(template: string, context: IDeviceContext): void {
    const html = template
      .replace(/{activeClass}/g, '')
      .replace(/{index}/g, 'Add')
      .replace(/{name}/g, 'Add Device');
    this.$tabs.append(html);
    const $addButton = this.$tabs.find('#deviceTabAdd');
    $addButton.on('click', () => {
      context.configuration.devices.push(this.getDefaultDeviceConfiguration(context.configSchema));
      this.renderDevicesSettings(context, context.configuration.devices.length - 1);
    });
  }

  private renderDeviceTabPanel(
    context: IDeviceContext,
    deviceConfiguration: PluginDeviceConfig,
    index: number,
    activeIndex: number,
    $tab: JQuery
  ): void {
    const tabPanelTemplate = `
    <div class="tab-pane container fade {activeClass} card card-body list-group-item" id="deviceTabPanel{index}">
      <button type="button" class="close pull-right" id="deviceTabPanelClose{index}">
        <span>Ã—</span>
        <span class="sr-only">Close</span>
      </button>
    </div>
  `;

    const activeTabPanelClass = index === activeIndex ? 'in show active' : '';
    const tabPanelHtml = tabPanelTemplate
      .replace(/{activeClass}/g, activeTabPanelClass)
      .replace(/{index}/g, index.toString());

    this.$tabPanels.append(tabPanelHtml);
    const $tabPanel = this.$tabPanels.find(`#deviceTabPanel${index}`);

    const deviceSchema = this.clone(context.configSchema.schema.properties.devices.items);
    const modelSchema = deviceSchema.properties.model;
    this.renderModelDropDown($tab, $tabPanel, index, context, deviceSchema, modelSchema!, deviceConfiguration);
    this.renderRemoveButton($tabPanel, index, context);
    this.renderForm($tab, index, activeIndex, context, deviceSchema, deviceConfiguration);
  }

  private renderModelDropDown(
    $tab: JQuery,
    $tabPanel: JQuery,
    index: number,
    context: IDeviceContext,
    deviceSchema: PluginConfigSchemaDevicesItems,
    modelSchema: PluginConfigSchemaEnum,
    deviceConfiguration: PluginDeviceConfig
  ) {
    this.componentRenderer.renderDropDown(
      $tabPanel,
      'model' + index,
      modelSchema,
      deviceConfiguration.model,
      newValue => {
        deviceConfiguration.model = newValue;
        this.renderForm($tab, index, index, context, deviceSchema, deviceConfiguration);
      }
    );
  }

  private renderRemoveButton($parent: JQuery, index: number, context: IDeviceContext) {
    const $removeButton = $parent.find(`#deviceTabPanelClose${index}`);
    $removeButton.on('click', () => {
      context.configuration.devices.splice(index, 1);
      this.renderDevicesSettings(context, index - 1);
    });
  }

  private renderForm(
    $tab: JQuery,
    index: number,
    activeIndex: number,
    context: IDeviceContext,
    deviceSchema: PluginConfigSchemaDevicesItems,
    deviceConfiguration: PluginDeviceConfig
  ): void {
    deviceSchema = this.clone(deviceSchema);
    delete deviceSchema.properties.model;
    context.hideDeviceSettingsPerModel[deviceConfiguration.model].forEach(propertyName => {
      delete deviceSchema.properties[propertyName];
    });

    $tab.off('click').on('click', () => {
      this.form?.end();
      this.form = context.homebridgeProvider.createForm({ schema: deviceSchema }, deviceConfiguration);
      this.form.onChange(async newDeviceConfiguration => {
        const oldName = context.configuration.devices[index].name;
        this.applyChanges(context.configuration.devices[index], newDeviceConfiguration, ['model']);
        await this.updatePluginConfig(context);

        if (newDeviceConfiguration.name !== oldName) {
          $tab.text(newDeviceConfiguration.name);
        }
      });
    });
    if (index === activeIndex) {
      $tab.trigger('click');
    }
  }

  private getDefaultDeviceConfiguration(configSchema: PluginConfigSchema): PluginDeviceConfig {
    const result = {} as PluginDeviceConfig;
    const properties = configSchema.schema.properties.devices.items.properties;
    Object.keys(properties)
      .filter(propertyName => !!properties[propertyName]?.default)
      .forEach(propertyName => {
        result[propertyName] = properties[propertyName]?.default;
      });
    return result;
  }

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  private applyChanges(target: any, source: any, readOnlyProperties: string[]): void {
    Object.assign(target, source);
    for (const key in target) {
      if (source[key] === undefined && !readOnlyProperties.includes(key)) {
        delete target[key];
      }
    }
  }
}
