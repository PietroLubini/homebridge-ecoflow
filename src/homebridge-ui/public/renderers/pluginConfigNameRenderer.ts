import { IContext } from '../interfaces/contracts';
import { PluginConfigRendererBase } from './pluginConfigRendererBase';

export class PluginConfigNameRenderer extends PluginConfigRendererBase {
  public render(context: IContext): void {
    this.componentRenderer.renderTextBox(
      $('#generalSettings'),
      'name',
      context.configSchema.schema.properties.name,
      context.configuration.name,
      async newValue => {
        context.configuration.name = newValue;
        await this.updatePluginConfig(context);
      }
    );
  }
}
