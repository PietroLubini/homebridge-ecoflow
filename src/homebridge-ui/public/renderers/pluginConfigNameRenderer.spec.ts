import { IContext } from '@ecoflow/homebridge-ui/public/interfaces/contracts';
import { IHomebridge, PluginConfig, PluginConfigSchema } from '@ecoflow/homebridge-ui/public/interfaces/homebridge';
import { ComponentRenderer } from '@ecoflow/homebridge-ui/public/renderers/componentRenderer';
import { PluginConfigNameRenderer } from '@ecoflow/homebridge-ui/public/renderers/pluginConfigNameRenderer';
import $ from 'jquery';

describe('PluginConfigNameRenderer', () => {
  let context: IContext;
  let homebridgeProviderMock: jest.Mocked<IHomebridge>;
  let configuration: PluginConfig;
  let configSchema: PluginConfigSchema;
  let generalSettingsElement: HTMLElement;
  let renderer: PluginConfigNameRenderer;

  beforeAll(() => {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    (global as any).$ = $;
  });

  afterAll(() => {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    (global as any).$ = undefined;
  });

  beforeEach(() => {
    homebridgeProviderMock = { updatePluginConfig: jest.fn() } as unknown as jest.Mocked<IHomebridge>;
    configuration = { name: 'name1', devices: [{ name: 'device1' }] } as PluginConfig;
    configSchema = { schema: { properties: { name: { title: 'Name Title' } } } } as PluginConfigSchema;
    context = { homebridgeProvider: homebridgeProviderMock, configuration, configSchema };

    generalSettingsElement = document.createElement('div');
    generalSettingsElement.id = 'generalSettings';
    document.body.replaceChildren(generalSettingsElement);

    renderer = new PluginConfigNameRenderer(new ComponentRenderer());
  });

  describe('render', () => {
    it('should render a textbox for name property when is called', () => {
      renderer.render(context);

      expect(generalSettingsElement).toMatchSnapshot();
    });

    it('should updatePluginConfig when value of textbox is changed', () => {
      renderer.render(context);
      const $control = $(document).find('#devicename');
      const newValue = 'name2';
      $control.val(newValue).trigger('change');

      expect(homebridgeProviderMock.updatePluginConfig).toHaveBeenCalledWith([
        { name: 'name2', devices: [{ name: 'device1' }] },
      ]);
    });
  });
});
