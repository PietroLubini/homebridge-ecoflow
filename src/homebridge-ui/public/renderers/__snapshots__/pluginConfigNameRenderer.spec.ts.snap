// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PluginConfigNameRenderer render should render a textbox for name property when is called 1`] = `
<div
  id="generalSettings"
>
  
    
  <div
    class="form-group"
  >
    
      
    <label
      class=""
      for="devicename"
    >
      Name Title
    </label>
    
      
    <input
      class="form-control"
      id="devicename"
      placeholder=""
      required="false"
      type="text"
    />
    
    
  </div>
</div>
`;
